### ISSUE TITLE
Stops working if there's only one workspace

### ISSUE THREAD

**Issue by henrythor** [bulletmark/libinput-gestures#31]
In Gnome 3.22 there's at first only one workspace, and another one doesn't become available until you run something in that workspace.

In those cases when you first launch Gnome and if you try the gesture, it will fail and you'll need to restart via libinput-gestures-setup to get them working again.

**Comment by bulletmark** [comment]
I don't know why you would think this is a bug? GNOME 3 has always worked like this. GNOME uses dynamic workspaces and only creates a new workspace when you have at least 1 window on all current workspaces. Experiment with the Super+PageUp/PageDown shortcuts to see this behaviour. You may be confused by the quirky behaviour of GNOME for the case of a single desktop when you create a window and then delete it then it will pop up the workspace indicator for 2 workspaces even though there is really only one workspace in that case. Either way, that is stock GNOME behaviour and restarting `libinput-gestures` makes no difference to anything.

**Comment by henrythor** [comment]
I'm not confused but maybe I'm bad at communicating.

If I gesture while there's 1 workspace, the thing monitoring libinput dies
and I can't gesture after I get another workspace.

That's my bug

17. nóv. 2016 22:09 skrifaði "Mark Blakeney" <notifications@github.com>:

**Comment by bulletmark** [comment]
What version are you using and what platform are you running on?

**Comment by bulletmark** [comment]
Also, please look in your journal and report here the error message at death, if any.

**Comment by henrythor** [comment]
libinput-gestures 2.10
libinput 1.5.0
Gnome 3.22.2
xf86-input-libinput 0.22.0
X.org 1.18.4

Antergos Linux

17. nóv. 2016 23:23 skrifaði "Mark Blakeney" <notifications@github.com>:

What version are you using and what platform are you running on?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/bulletmark/libinput-gestures/issues/31#issuecomment-261401773>,
or mute the thread

**Comment by bulletmark** [comment]
I don't know Antergos very well but your versions seem to be the same as mine on stock Arch (which I know Antergos is based on). Please look through your journal for error messages when this problem happens.

**Comment by henrythor** [comment]
It's easily reproducible (for me) so this is what I get when I first get into Gnome and try to switch:
```
Nov 17 23:37:53 tempest libinput-gestures.desktop[726]: Cannot get number of desktops properties. (_NET_NUMBER_OF_DESKTOPS or _WIN_WORKSPACE_COUNT)
Nov 17 23:37:53 tempest libinput-gestures.desktop[726]: Command '['wmctrl', '-d']' returned non-zero exit status 1
```

Then I open Chromium to type this, and now I have a second workspace I can switch to via Super-PgDn, but I can't use the gesture to get there. But there's nothing landing in the journal and /usr/bin/libinput-gestures isn't running:

```
➜  ~ ps auxw|grep libinput
henry     2427  0.0  0.0  10864  2216 pts/0    S+   23:43   0:00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn libinput
➜  ~ libinput-gestures-setup stop
➜  ~ libinput-gestures-setup start
libinput-gestures started.
➜  ~ ps auxw|grep libinput        
henry     2466  0.3  0.0  40788 11276 pts/0    S    23:43   0:00 python3 /usr/bin/libinput-gestures
henry     2473  0.0  0.0  55364  2408 pts/0    S    23:43   0:00 libinput-debug-events --device /dev/input/event7
henry     2485  0.0  0.0  10864  2216 pts/0    S+   23:43   0:00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn libinput
➜  ~ 
```

**Comment by henrythor** [comment]
(And after restarting, it works)

**Comment by bulletmark** [comment]
Yes, it dies due to that error message and after it dies it is not going to work. That error message is pretty important! Are you running using Wayland or Xorg?

**Comment by henrythor** [comment]
Xorg

**Comment by bulletmark** [comment]
OK, I can reproduce this on Xorg for the case you report. Leave it with me to fix. It seems `wmctrl` has some issue when only 1 desktop is present, and then only sometimes.

Thanks for reporting this bug.

**Comment by henrythor** [comment]
I tried to debug some more, but I don't know these systems.. Will leave this if it helps:
```
➜  ~ libinput-gestures-setup stop
libinput-debug-events stopped.
libinput-gestures stopped.
➜  ~ libinput-gestures -v
libinput-gestures: monitoring device /dev/input/event8 for 6 gestures
libinput-gestures: SWIPE up 3 [82.56, -1040.0199999999998]
   _internal ws_up
Cannot get number of desktops properties. (_NET_NUMBER_OF_DESKTOPS or _WIN_WORKSPACE_COUNT)
Command '['wmctrl', '-d']' returned non-zero exit status 1
➜  ~ wmctrl -d
Cannot get number of desktops properties. (_NET_NUMBER_OF_DESKTOPS or _WIN_WORKSPACE_COUNT)
➜  ~ echo $? 
1
➜  ~ wmctrl -d -v
envir_utf8: 1
Invalid type of _NET_NUMBER_OF_DESKTOPS property.
Invalid type of _WIN_WORKSPACE_COUNT property.
Cannot get number of desktops properties. (_NET_NUMBER_OF_DESKTOPS or _WIN_WORKSPACE_COUNT)
➜  ~ xprop -root | grep "_NET_NUMBER_OF_DESKTOPS"
_NET_SUPPORTED(ATOM) = _NET_WM_NAME, _NET_CLOSE_WINDOW, _NET_WM_STATE, _NET_WM_STATE_SHADED, _NET_WM_STATE_MAXIMIZED_HORZ, _NET_WM_STATE_MAXIMIZED_VERT, _NET_WM_DESKTOP, _NET_NUMBER_OF_DESKTOPS, _NET_CURRENT_DESKTOP, _NET_WM_WINDOW_TYPE, _NET_WM_WINDOW_TYPE_DESKTOP, _NET_WM_WINDOW_TYPE_DOCK, _NET_WM_WINDOW_TYPE_TOOLBAR, _NET_WM_WINDOW_TYPE_MENU, _NET_WM_WINDOW_TYPE_UTILITY, _NET_WM_WINDOW_TYPE_SPLASH, _NET_WM_WINDOW_TYPE_DIALOG, _NET_WM_WINDOW_TYPE_DROPDOWN_MENU, _NET_WM_WINDOW_TYPE_POPUP_MENU, _NET_WM_WINDOW_TYPE_TOOLTIP, _NET_WM_WINDOW_TYPE_NOTIFICATION, _NET_WM_WINDOW_TYPE_COMBO, _NET_WM_WINDOW_TYPE_DND, _NET_WM_WINDOW_TYPE_NORMAL, _NET_WM_STATE_MODAL, _NET_CLIENT_LIST, _NET_CLIENT_LIST_STACKING, _NET_WM_STATE_SKIP_TASKBAR, _NET_WM_STATE_SKIP_PAGER, _NET_WM_ICON, _NET_WM_ICON_GEOMETRY, _NET_WM_MOVERESIZE, _NET_ACTIVE_WINDOW, _NET_WM_STRUT, _NET_WM_STATE_HIDDEN, _NET_WM_STATE_FULLSCREEN, _NET_WM_PING, _NET_WM_PID, _NET_WORKAREA, _NET_SHOWING_DESKTOP, _NET_DESKTOP_LAYOUT, _NET_DESKTOP_NAMES, _NET_WM_ALLOWED_ACTIONS, _NET_WM_ACTION_MOVE, _NET_WM_ACTION_RESIZE, _NET_WM_ACTION_SHADE, _NET_WM_ACTION_STICK, _NET_WM_ACTION_MAXIMIZE_HORZ, _NET_WM_ACTION_MAXIMIZE_VERT, _NET_WM_ACTION_CHANGE_DESKTOP, _NET_WM_ACTION_CLOSE, _NET_WM_STATE_ABOVE, _NET_WM_STATE_BELOW, _NET_STARTUP_ID, _NET_WM_STRUT_PARTIAL, _NET_WM_ACTION_FULLSCREEN, _NET_WM_ACTION_MINIMIZE, _NET_FRAME_EXTENTS, _NET_REQUEST_FRAME_EXTENTS, _NET_WM_USER_TIME, _NET_WM_STATE_DEMANDS_ATTENTION, _NET_MOVERESIZE_WINDOW, _NET_DESKTOP_GEOMETRY, _NET_DESKTOP_VIEWPORT, _NET_WM_USER_TIME_WINDOW, _NET_WM_ACTION_ABOVE, _NET_WM_ACTION_BELOW, _NET_WM_STATE_STICKY, _NET_WM_FULLSCREEN_MONITORS, _NET_WM_STATE_FOCUSED, _NET_WM_BYPASS_COMPOSITOR, _NET_WM_OPAQUE_REGION, _NET_WM_FRAME_DRAWN, _NET_WM_FRAME_TIMINGS, _NET_WM_WINDOW_OPACITY, _GTK_FRAME_EXTENTS, _GTK_SHOW_WINDOW_MENU
➜  ~ xprop -root _NET_NUMBER_OF_DESKTOPS    
_NET_NUMBER_OF_DESKTOPS:  not found.
➜  ~ xprop -root _WIN_WORKSPACE_COUNT   
_WIN_WORKSPACE_COUNT:  not found.
➜  ~ 
```

**Comment by bulletmark** [comment]
Yes, I can see what is happening. It is probably a bug in GNOME but likely affects little else other than `wmctrl` so I will just intercept the error condition.

**Comment by henrythor** [comment]
Cool. Thanks for looking at this.

**Comment by bulletmark** [comment]
Fixed in the commit above and released as v2.11 to AUR.

**Comment by henrythor** [comment]
Installed via AUR on my end and works perfectly.
